internal/core/software_renderer.rs

```diff
@@ -454,6 +478,34 @@ impl SoftwareRenderer {
     /// Returns the physical dirty region for this frame, excluding the extra_draw_region,
     /// in the window frame of reference. It is affected by the screen rotation.
     pub fn render(&self, buffer: &mut [impl TargetPixel], pixel_stride: usize) -> PhysicalRegion {
+        self.render_buffer_impl(&mut TargetPixelSlice { data: buffer, pixel_stride })
+    }
+
+    /// Render the window to the given frame buffer.
+    ///
+    /// The renderer uses a cache internally and will only render the part of the window
+    /// which are dirty. The `extra_draw_region` is an extra region which will also
+    /// be rendered. (eg: the previous dirty region in case of double buffering)
+    /// This function returns the region that was rendered.
+    ///
+    /// The buffer's line slices need to be wide enough to if the `width` of the screen and the line count the `height`,
+    /// or the `height` and `width` swapped if the screen is rotated by 90°.
+    ///
+    /// Returns the physical dirty region for this frame, excluding the extra_draw_region,
+    /// in the window frame of reference. It is affected by the screen rotation.
+    #[cfg(feature = "experimental")]
+    pub fn render_into_buffer(&self, buffer: &mut impl TargetPixelBuffer) -> PhysicalRegion {
+        self.render_buffer_impl(buffer)
+    }
+
+    fn render_buffer_impl(
+        &self,
+        buffer: &mut impl target_pixel_buffer::TargetPixelBuffer,
+    ) -> PhysicalRegion {
+        let pixel_stride = buffer.line_slice(0).len();
```

Comment:
That's not technically a stride.

Since this is only used to assert here that the buffer is big enough, I think it'd be fine to move that assert back to the render() function.
I'm not sure the case where the window_item is None is actually not dead code anyway. But we probably can get the size from the window's size

Reply:
Good point!

Reply:
One second thought, I see value in having this in this function instead of just `render()`, as even a provided `TargetPixelBuffer` could have wrong dimensions. But I'll rename the variable, that's indeed not the stride.